liquidity-amo/contracts/V3AMO.sol

```diff
@@ -0,0 +1,578 @@
+        if (msg.sender != pool) revert UntrustedCaller(msg.sender);
+
+        (uint256 boostOwed, uint256 usdOwed) = sortAmounts(amount0Owed, amount1Owed);
+        uint256 boostAmount = (toBoostAmount(usdOwed) * boostMultiplier) / FACTOR;
+        if (boostAmount < boostOwed) revert InvalidOwed();
+
+        IERC20Upgradeable(usd).safeTransfer(pool, usdOwed);
+        IMinter(boostMinter).protocolMint(pool, boostOwed);
+    }
+
+    ////////////////////////// AMO_ROLE ACTIONS //////////////////////////
+    function _mintAndSellBoost(
+        uint256 boostAmount
+    ) internal override returns (uint256 boostAmountIn, uint256 usdAmountOut) {
+        // Mint BOOST and execute the swap BOOST->USD
+        // The swap is executed at the targetSqrtPriceX96
+        (int256 amount0, int256 amount1) = IUniswapV3Pool(pool).swap(
+            address(this),
+            boost < usd, // zeroForOne
+            int256(boostAmount), // Amount of BOOST tokens being swapped
+            targetSqrtPriceX96, // The target square root price
+            abi.encode(SwapType.SELL)
+        );
+
+        (int256 boostDelta, int256 usdDelta) = sortAmounts(amount0, amount1);
+        boostAmountIn = uint256(boostDelta); // BOOST tokens used in the swap
+        usdAmountOut = uint256(-usdDelta); // USD tokens received from the swap
+
+        emit MintSell(boostAmountIn, usdAmountOut);
+    }
+
+    /**
+     * @notice Calculates the liquidity required to match a given USD amount.
+     * @dev This function ensures accurate liquidity estimation by using the current pool state and tick bounds.
+     *      It avoids inaccuracies by leveraging the Uniswap V3 price and liquidity formulas.
+     * @param usdAmount The amount of USD for which the corresponding liquidity is to be calculated.
+     * @return liquidity The calculated liquidity amount corresponding to the given USD amount.
+     */
+    function _getLiquidityForUsdAmount(uint256 usdAmount) internal view returns (uint256 liquidity) {
+        // Step 1: Fetch the current price and pool data
+        uint160 sqrtRatioX96 = _getSqrtPriceX96();
+
+        // Step 2: Fetch the price bounds corresponding to the tickLower and tickUpper
+        uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(tickLower);
+        uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(tickUpper);
+
+        // Step 3: Sort amounts to determine the maximum amounts of token0 (BOOST) and token1 (USD)
+        (uint256 amount0, uint256 amount1) = sortAmounts(type(uint128).max, usdAmount);
+
+        // Step 4: Use the Uniswap V3 LiquidityAmounts library to calculate liquidity
+        liquidity = uint256(
+            LiquidityAmounts.getLiquidityForAmounts(
+                sqrtRatioX96, // Current pool price
+                sqrtRatioAX96, // Lower bound price
+                sqrtRatioBX96, // Upper bound price
+                amount0, // Maximum BOOST
+                amount1 // Maximum USD
```

Comment:
Nit: the code comment is wrong here. amount0 may not be Boost in case boost > usd.